### PR TITLE
Join single-line comments with J.

### LINKED ISSUE #8565: Remove comment token when joining two comment lines
Example of current behaviour:
```hs
-- Foo
-- Bar
```
turns into
```hs
-- Foo -- Bar
```
---
Example of desired behaviour:
```hs
-- Foo
-- Bar
```
turns into
```hs
-- Foo Bar
```
---
Since helix is aware of the comment tokens, it should be possible to detect if the user is joining two comment lines, and remove the comment token (including succeeding whitespace).

A subtlety here is that this should only ever happen if both lines are comments.

Any thoughts on this feature?

I could try implementing this myself sometime soon if I find the time.

### PR DESCRIPTION
**Pull request by rhogenson**
Fixes #8565.

### CODE REVIEW

File: helix-term/src/commands.rs [hunk 1]
```diff
@@ -4626,6 +4626,15 @@ fn join_selections_impl(cx: &mut Context, select_space: bool) {
     let text = doc.text();
     let slice = text.slice(..);
 
+    let comment_prefixes = doc
```
Comment (the-mikedavis) [inline helix-term/src/commands.rs:4629]: minor naming nit: let's call these `comment_tokens` to match the language we use for configuration

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -4626,6 +4626,15 @@ fn join_selections_impl(cx: &mut Context, select_space: bool) {
...
+        .language_config()
+        .and_then(|config| config.comment_tokens.as_ref())
+        .map(|pfxs| &pfxs[..])
```
Comment (the-mikedavis) [inline helix-term/src/commands.rs:4632]:
```suggestion
        .and_then(|config| config.comment_tokens.as_deref())
```

File: helix-term/src/commands.rs [hunk 3]
```diff
@@ -4641,6 +4650,14 @@ fn join_selections_impl(cx: &mut Context, select_space: bool) {
             let start = line_end_char_index(&slice, line);
             let mut end = text.line_to_char(line + 1);
             end = skip_while(slice, end, |ch| matches!(ch, ' ' | '\t')).unwrap_or(end);
+            let slice_from_end = slice.slice(end..);
+            for pfx in comment_prefixes.iter() {
+                if slice_from_end.starts_with(pfx) {
+                    end += pfx.len();
+                    break;
+                }
+            }
```
Comment (the-mikedavis) [inline helix-term/src/commands.rs:4659]:
```suggestion
            if let Some(token) = comment_tokens
                .iter()
                .find(|token| slice_from_end.starts_with(token))
            {
                end += token.len();
            }
```

File: helix-term/src/commands.rs [hunk 4]
```diff
@@ -4641,6 +4650,14 @@ fn join_selections_impl(cx: &mut Context, select_space: bool) {
             let start = line_end_char_index(&slice, line);
             let mut end = text.line_to_char(line + 1);
             end = skip_while(slice, end, |ch| matches!(ch, ' ' | '\t')).unwrap_or(end);
+            let slice_from_end = slice.slice(end..);
+            for pfx in comment_prefixes.iter() {
+                if slice_from_end.starts_with(pfx) {
+                    end += pfx.len();
```
Comment (the-mikedavis) [inline helix-term/src/commands.rs:4656]: This is mixing up byte and character indexing: `end` is a character index into the rope and the token's `len()` is the number of bytes in its UTF-8 representation. It won't make a difference in the common case (comment token is ASCII) but if you try a comment token that takes more than one byte in UTF-8 `end` will be greater than it should. Instead here we have to use `token.chars().count()`.

**Review by the-mikedavis — COMMENTED** [review]
I reflow my comments a lot so I'm definitely looking forward to this! I left some minor comments.

There's an issue with J being used on the line before a comment. It strips the comment token as if it were joining two comment lines. We'll need to detect if the line we're joining to is a line comment (ideally of the same kind) to avoid this

**Comment by rhogenson** [comment]
Thank you for the detailed comments! I updated the logic to only delete a token if it matches the comment token on the previous line.

**Review by pascalkuthe — APPROVED** [review]
thanks for adding some tests :+1:

**Review by archseer — APPROVED** [review]

**Pull request merged**